### PR TITLE
pilot-core: changes for flying practice

### PR DESCRIPTION
**Pull request by niewim19**
## Description

Following changes were made to achieve working quadrocopter in real scenario flight

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: quadcontrol/control.c [hunk 1]
```diff
@@ -420,7 +496,7 @@ static void quad_rcbusHandler(const rcbus_msg_t *msg)
 	/* Manual Disarm: SWA == MIN, SWB == MIN, SWC == MIN, SWD == MIN, Throttle == 0 && mode_rc */
 	if (msg->channels[RC_SWA_CH] <= minTriggerVal && msg->channels[RC_SWB_CH] <= minTriggerVal
 			&& msg->channels[RC_SWC_CH] <= minTriggerVal && msg->channels[RC_SWD_CH] <= minTriggerVal
-			&& msg->channels[RC_LEFT_VSTICK_CH] <= minTriggerVal && quad_common.mode == mode_rc) {
+			&& msg->channels[RC_LEFT_VSTICK_CH] <= minTriggerVal && quad_common.currFlight == flight_idle) {
```
[the remark this answers is not in the document]
Comment (niewim19) [reply quadcontrol/control.c:499]: I wouldn`t implement this suggestion as it makes code unreadable

File: scripts/quadLive.py [hunk 2]
```diff
@@ -189,10 +189,11 @@ def main():
     cmdParse.add_argument("port", default="/dev/ttyUSB0")
     cmdParse.add_argument("-c", "--cut", required=False, default=200, type=int)
     cmdParse.add_argument("-p", "--pwm", required=False, default=0.15, type=float)
+    cmdParse.add_argument("-b", "--baud", required=False, default=115200, type=int)
     args = cmdParse.parse_args()
 
     try:
-        ser = serial.Serial(args.port, baudrate=115200, timeout=1)
+        ser = serial.Serial(args.port, baudrate=args.baud, timeout=1)
```
[the remark this answers is not in the document]
Comment (niewim19) [reply scripts/quadLive.py:196]: it should be like that

File: quadcontrol/control.c [hunk 3]
```diff
@@ -273,15 +285,55 @@ static int quad_landing(const flight_mode_t *mode)
...
+	mutexUnlock(quad_common.rcbusLock);
+
+	if (stickSWD >= highTrigTresh || rcThrottle >= lowTrigTresh) {
+		fprintf(stderr, "Wrong initial transmitter setup. Aborting...\n");
+		return -1;
+	}
+
+	throttle = hoverThrtlEst * ((float)rcThrottle / (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) + 0.5); /* throttle now in range (0.5, 1.5) of hover throttle */
+
 	now = quad_timeMsGet();
```
Comment (HBuczynski) [inline quadcontrol/control.c:337]: You can remove it.

File: quadcontrol/control.c [hunk 4]
```diff
@@ -273,15 +285,55 @@ static int quad_landing(const flight_mode_t *mode)
...
 
+	/* Read RC packet for variable initialization purposes */
+	mutexLock(quad_common.rcbusLock);
+	{
+		stickSWD = quad_common.rcChannels[RC_SWD_CH];
+		rcRoll = quad_common.rcChannels[RC_RIGHT_HSTICK_CH];
+		rcPitch = quad_common.rcChannels[RC_RIGHT_VSTICK_CH];
+		rcYaw = quad_common.rcChannels[RC_LEFT_HSTICK_CH];
+		rcThrottle = quad_common.rcChannels[RC_LEFT_VSTICK_CH];
+	}
```
[the remark this answers is not in the document]
Comment (niewim19) [reply quadcontrol/control.c]: To emphasize mutex-locked part of code. Removed as requested.

Btw. our phoenix-rtos-doc coding convention does not restrict or even mention additional code block usage, or did I miss something?

File: quadcontrol/control.c [hunk 5]
```diff
@@ -438,11 +516,9 @@ static void quad_rcbusHandler(const rcbus_msg_t *msg)
...
 
-	if (quad_common.currFlight == flight_manual) {
-		mutexLock(quad_common.rcbusLock);
-		memcpy(quad_common.rcChannels, msg->channels, sizeof(quad_common.rcChannels));
-		mutexUnlock(quad_common.rcbusLock);
-	}
+	mutexLock(quad_common.rcbusLock);
+	memcpy(quad_common.rcChannels, msg->channels, sizeof(quad_common.rcChannels));
+	mutexUnlock(quad_common.rcbusLock);
 }
```
[the remark this answers is not in the document]
Comment (niewim19) [reply quadcontrol/control.c]: If clause removal was recerted. Now it is copied only in `mode_rc`.

File: quadcontrol/control.c (hunk 3 above)
Comment (niewim19) [reply quadcontrol/control.c:336]: But then we risk uninitialized `throttle` variable if I switch SWD high during engines arming (by accident lets say). With this line we ensure that we start with throttle variable equal to that of RC transmitter. Also removing this line causes compilation warning of possible uninitialized variable.

Comment (HBuczynski) [reply quadcontrol/control.c:336]: Ups, my mistake. The comment was to line 337.

**Review by HBuczynski — CHANGES_REQUESTED** [review]

File: quadcontrol/control.c [hunk 6]
```diff
@@ -273,38 +284,89 @@ static int quad_landing(const flight_mode_t *mode)
...
+* - control over roll/pitch
+* - throttle is left as was in STABILIZE mode
+*/
 static int quad_manual(void)
 {
-	float throttle;
-	time_t now, lastLog = 0;
-	int32_t alt, roll, pitch, yaw;
+	const static uint16_t highTrigTresh = 95 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 100 + MIN_CHANNEL_VALUE;
+	const static uint16_t lowTrigTresh = 5 * (MAX_CHANNEL_VALUE - MIN_CHANNEL_VALUE) / 100 + MIN_CHANNEL_VALUE;
```
Comment (HBuczynski) [inline quadcontrol/control.c:303]: These lines are copied from `quad_rcbusHandler`. You can create defines and update also the `quad_rcbusHandler`

**Comment by niewim19** [comment]
All review remarks were implemented in old commits.

After last review two commits were added:
 - mma motors order change so that drone software front is in the same direction as drone hardware front
 - ekf development config read was added, as it was added each time manually for testing purposes

File: libs/mctl/mctl.c [hunk 7]
```diff
@@ -155,23 +156,41 @@ bool mctl_isArmed(void)
 }
 
 
+/* 
+* Disarm procedure has two steps:
+* 1) Setting the motor speed to 0, 
+* 2) Disarming the motors by disabling PWM generation
+*/
```
Comment (HBuczynski) [inline libs/mctl/mctl.c:163]: The comments inside a function are enough.

File: quadcontrol/control.c [hunk 8]
```diff
@@ -43,12 +44,18 @@
...
 #define PID_NUMBERS      4
 
-#define ANGLE_THRESHOLD (M_PI / 6)
+#define HOVER_THROTTLE 0.27
+
+#define ANGLE_THRESHOLD (M_PI / 4)
 #define RAD2DEG         ((float)180.0 / M_PI)
 #define DEG2RAD         0.0174532925f
-#define ANGLE_HOLD      INT32_MAX
+#define RC_IIR_LATENCY  3
```
Comment (HBuczynski) [inline quadcontrol/control.c:52]: It is used only in one function. I would not make it a global define.
You can define it as `static const` inside `quad_iir`.

File: quadcontrol/control.c [hunk 9]
```diff
@@ -208,26 +221,28 @@ static int quad_takeoff(const flight_mode_t *mode)
...
-	time_t now, end, lastLog;
+	quad_att_t att;
+	time_t now, end;
+	ekf_state_t measure;
+
+	ekf_stateGet(&measure);
+
+	att.yaw = measure.yaw;
+	att.pitch = 0;
+	att.roll = 0;
```
Comment (HBuczynski) [inline quadcontrol/control.c:232]:
```suggestion
	quad_att_t att = { 0 };
	time_t now, end;
	ekf_state_t measure;

	ekf_stateGet(&measure);

	att.yaw = measure.yaw;
```

File: quadcontrol/control.c [hunk 10]
```diff
@@ -171,8 +182,16 @@ static int quad_motorsCtrl(float throttle, int32_t alt, int32_t roll, int32_t pi
...
 	float throttle, coeff;
-	time_t spoolStart, spoolEnd, now, lastLog = 0;
+	time_t spoolStart, spoolEnd, now;
+	ekf_state_t measure;
+
+	ekf_stateGet(&measure);
+
+	att.yaw = measure.yaw;
+	att.pitch = 0;
+	att.roll = 0;
```
Comment (HBuczynski) [inline quadcontrol/control.c:194]:
```suggestion
        quad_att_t att = { 0 };
	time_t spoolStart, spoolEnd, now;
	ekf_state_t measure;

	ekf_stateGet(&measure);

	att.yaw = measure.yaw;
```

File: quadcontrol/control.c [hunk 11]
```diff
@@ -237,11 +252,26 @@ static int quad_hover(const flight_mode_t *mode)
 	return 0;
 }
 
+/* Infinite impulse response filter */
+static int32_t quad_iir(int32_t curr, int32_t new)
```
Comment (HBuczynski) [inline quadcontrol/control.c:256]:
```suggestion
static inline int32_t quad_iir(int32_t curr, int32_t new)
```

File: quadcontrol/control.c [hunk 12]
```diff
@@ -237,11 +252,26 @@ static int quad_hover(const flight_mode_t *mode)
...
+	ekf_state_t measure;
 	float throttle, coeff;
-	time_t spoolStart, spoolEnd, now, lastLog = 0;
+	time_t spoolStart, spoolEnd, now;
+
+	ekf_stateGet(&measure);
+
+	att.yaw = measure.yaw;
+	att.pitch = 0;
+	att.roll = 0;
```
Comment (HBuczynski) [inline quadcontrol/control.c:273]: The same case as above.

**Review by HBuczynski — APPROVED** [review]

**Pull request merged**